ember-server/shared/src/main/scala/org/http4s/ember/server/internal/WebSocketHelpers.scala

```diff
@@ -193,15 +194,29 @@ object WebSocketHelpers {
       rest.pull.uncons.flatMap {
         case Some((chunk, next)) =>
           val buffer = acc ++ chunk.toArray[Byte]
-          val byteBuffer = ByteBuffer.wrap(buffer)
+          var byteBuffer = ByteBuffer.wrap(buffer)
           Pull
-            .eval(F.delay(nonClientTranscoder.bufferToFrame(byteBuffer)))
-            .flatMap { value =>
+            .eval(F.delay {
+              // A single chunk might contain multiple frames
+              // but `bufferToFrame` decodes at most one, so
+              // we call it repeatedly until all frames in the
+              // buffer are decoded.
+              val frames = ArrayBuffer.empty[WebSocketFrame]
+              var frame = nonClientTranscoder.bufferToFrame(byteBuffer)
+              while (frame != null) {
+                frames += frame
+                // We need to slice b/c `bufferToFrame` does absolute reads.
+                byteBuffer = byteBuffer.slice()
+                frame = nonClientTranscoder.bufferToFrame(byteBuffer)
+              }
+              Chunk.array(frames.toArray)
+            })
```

Comment:
I wonder if we should write this as a tail-recursion, particularly in `F[_]`. This way if there is a big chunk containing multiple frames the runtime can auto-cede during parsing and avoid hogging the thread, which is better for fairness.

Of course there is a performance trade-off here, so ideally someone else should weigh in :) mostly I just wanted to point this out.

Reply:
This should be tail recursion. This offers relatively low performance benefit by comparison to putting a var in a position like this. 

Functional closures. But this closure is in fact cheating with a mutable in a non-pure position. 

So, this can move to a function that is pure OR this should be tail-recursive. But mutability should not be in place like this.